From f68b2ff450218b849a9405e171b4e4d228873881 Mon Sep 17 00:00:00 2001
From: Patrick Delaunay <patrick.delaunay@foss.st.com>
Date: Mon, 2 Dec 2024 17:55:51 +0100
Subject: [PATCH 796/841] feat(stm32mp1): add crc in low power mode context

Add support of MAILBOX_MAGIC_V4 with crc and header size.

TF-A BL2 with check integrity of the low power mode context with CRC
on the OP-TEE context size for compatibility issue.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git 3edf4e35fd92f1840a4528dfb7f76d0e8abe7fb1 feat(stm32mp1): add crc in low power mode context]
Signed-off-by: Patrick Delaunay <patrick.delaunay@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/423285
Change-Id: I2ecf01ded9c6516629f15c22996b3b7271779458
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/437069
Reviewed-by: Maxime MERE <maxime.mere@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
---
 plat/st/stm32mp1/bl2_plat_setup.c   |  4 +--
 plat/st/stm32mp1/stm32mp1_context.c | 40 ++++++++++++++++++++++++++++-
 2 files changed, 40 insertions(+), 4 deletions(-)

diff --git a/plat/st/stm32mp1/bl2_plat_setup.c b/plat/st/stm32mp1/bl2_plat_setup.c
index 6d231d77a..05e525dc5 100644
--- a/plat/st/stm32mp1/bl2_plat_setup.c
+++ b/plat/st/stm32mp1/bl2_plat_setup.c
@@ -726,9 +726,7 @@ int bl2_plat_handle_post_image_load(unsigned int image_id)
 			bl_mem_params->ep_info.args.arg0 = 0;
 		}
 
-		if (bl_mem_params->ep_info.pc >= STM32MP_DDR_BASE) {
-			stm32_context_save_bl2_param();
-		}
+		stm32_context_save_bl2_param();
 		break;
 
 	case BL33_IMAGE_ID:
diff --git a/plat/st/stm32mp1/stm32mp1_context.c b/plat/st/stm32mp1/stm32mp1_context.c
index 1c89703cb..12f3aaf65 100644
--- a/plat/st/stm32mp1/stm32mp1_context.c
+++ b/plat/st/stm32mp1/stm32mp1_context.c
@@ -8,6 +8,7 @@
 
 #include <arch_helpers.h>
 #include <common/debug.h>
+#include <common/tf_crc32.h>
 #include <drivers/clk.h>
 #include <drivers/st/stm32mp1_ddr_helpers.h>
 #include <drivers/st/stm32mp1_ddr_regs.h>
@@ -32,10 +33,14 @@
  * MAILBOX_MAGIC_V3:
  * Context provides V2 content, low power entry point, BL2 code start, end and
  * BL2_END (102 bytes). And, only for STM32MP13, adds MCE master key (16 bytes).
+ *
+ * MAILBOX_MAGIC_V4:
+ * Context provides V3 content, size and CRC protection.
  */
 #define MAILBOX_MAGIC_V1		(0x0001U << 16U)
 #define MAILBOX_MAGIC_V2		(0x0002U << 16U)
 #define MAILBOX_MAGIC_V3		(0x0003U << 16U)
+#define MAILBOX_MAGIC_V4		(0x0004U << 16U)
 #define MAILBOX_MAGIC			(MAILBOX_MAGIC_V3 | \
 					 TRAINING_AREA_SIZE)
 
@@ -65,6 +70,8 @@ struct backup_data_s {
 	uint8_t mce_mkey[MCE_KEY_SIZE_IN_BYTES];
 	struct stm32_mce_region_s mce_regions[MCE_IP_MAX_REGION_NB];
 #endif
+	uint32_t size;
+	uint32_t crc_32;
 };
 
 uint32_t stm32_pm_get_optee_ep(void)
@@ -83,6 +90,7 @@ uint32_t stm32_pm_get_optee_ep(void)
 	case MAILBOX_MAGIC_V1:
 	case MAILBOX_MAGIC_V2:
 	case MAILBOX_MAGIC_V3:
+	case MAILBOX_MAGIC_V4:
 		if (MAGIC_AREA_SIZE(backup_data->magic) != TRAINING_AREA_SIZE) {
 			panic();
 		}
@@ -131,7 +139,7 @@ void stm32_context_save_bl2_param(void)
 	backup_data->bl2_code_base = BL_CODE_BASE;
 	backup_data->bl2_code_end = BL_CODE_END;
 	backup_data->bl2_end = BL2_END;
-	backup_data->magic = MAILBOX_MAGIC_V3;
+	backup_data->magic = MAILBOX_MAGIC_V4;
 	backup_data->zq0cr0_zdata = ddr_get_io_calibration_val();
 
 	clk_disable(BKPSRAM);
@@ -155,9 +163,23 @@ uint32_t stm32_get_zdata_from_context(void)
 	return zdata;
 }
 
+static uint32_t stm32mp_pm_crc_check(struct backup_data_s *data)
+{
+	uint32_t saved_crc;
+	uint32_t crc;
+
+	saved_crc = data->crc_32;
+	data->crc_32 = 0U;
+	crc =  tf_crc32(0U, (unsigned char *)data, data->size);
+	data->crc_32 = saved_crc;
+
+	return crc;
+}
+
 bool stm32_pm_context_is_valid(void)
 {
 	struct backup_data_s *backup_data;
+	uint32_t crc_32 = 0U;
 	bool ret;
 
 	clk_enable(BKPSRAM);
@@ -170,6 +192,22 @@ bool stm32_pm_context_is_valid(void)
 	case MAILBOX_MAGIC_V3:
 		ret = true;
 		break;
+	case MAILBOX_MAGIC_V4:
+		if (backup_data->size < sizeof(*backup_data)) {
+			ret = false;
+		} else {
+			crc_32 = stm32mp_pm_crc_check(backup_data);
+			ret = (crc_32 == backup_data->crc_32);
+		}
+		if (!ret) {
+			ERROR("pm context invalid (crc=%x vs %x, size=%u vs %u)\n",
+			      backup_data->crc_32, crc_32, backup_data->size, sizeof(*backup_data));
+			backup_data->magic = 0U;
+		}
+		VERBOSE("%s(crc=%x vs %x, size=%u vs %u, magic=%x vs %x) = %d\n", __func__,
+		        backup_data->crc_32, crc_32, backup_data->size, sizeof(*backup_data),
+		        backup_data->magic, MAILBOX_MAGIC_V4, ret);
+		break;
 	default:
 		ret = false;
 		break;
-- 
2.47.2

